### ISSUE TITLE
3.x Flowable#sample may complete on an interrupted thread

### ISSUE THREAD

**Issue by feneuilflo** [ReactiveX/RxJava#7201]
Hi !

Using version 3.0.10, and the following code:
```java
int i = 0;
while (true) {
        System.out.println("i = " + i++);
        PublishSubject<Long> subj = PublishSubject.create();
        subj.toFlowable(BackpressureStrategy.ERROR)
	        .sample(10, TimeUnit.MILLISECONDS)
	        .doFinally(() -> {
		        System.out.println(String.format("Current thread %s is interrupted: %b", 
                                        Thread.currentThread().getName(),
				        Thread.currentThread().isInterrupted())
                        );
		        if (Thread.currentThread().isInterrupted()) {
			        System.exit(1);
		        }
	        })
	        .subscribe();
        subj.onNext(1L);
        
        Observable.timer(10, TimeUnit.MILLISECONDS)
	        .blockingSubscribe(any -> subj.onComplete());
}
```
the loop ends after a few iterations (usually between 250 and 300 on my device).

The first iterations give: `Current thread main is interrupted: false`.
And the last: `Current thread RxComputationThreadPool-7 is interrupted: true`.

I quite understand why I get this result, but I'm still surprised by it. Is it an known/accepted/intended behavior ?

NB 1: Using `buffer` unstead of `sample` gives the same result.
NB 2: Changing `sample` scheduler to `Schedulers.from(Executors.newSingleThreadExecutor(...))` causes the loop to never end.
NB 3: First observed with RxJava2, with no change in behavior.

**Comment by akarnokd** [comment]
This is possible behavior because your completion races with the sampling. When the upstream completes, the timer for the sampler thread is cancelled which interrupts it if it happens to run.

If you want to avoid interruption, run the `sample` on a non-interruptible scheduler via [Schedulers.from(Executor, boolean)](http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/schedulers/Schedulers.html#from-java.util.concurrent.Executor-boolean-).

**Comment by feneuilflo** [comment]
Thanks for the explanation and the work-around.

I have been doing a few tests with `Schedulers.from(Executor, boolean)`.
Using `Executors.newSingleThreadExecutor()` has given the expected result (ie an interrupted thread with a value `true` for the boolean, no interruption otherwise).
Using `Executors.newSingleThreadScheduledExecutor()` has not. In both case my code ended with an interruption.
Is there something I have been missing ?

**Comment by akarnokd** [comment]
I don't understand what you are saying. Are you saying you tried my suggestion with those executors and still got the interruption in doFinally?

**Comment by feneuilflo** [comment]
That's it.
I ran the code provided in my first post, with only one modification: I specified a scheduler for the `sample` operator.

I tried the following schedulers:

`Schedulers.from(Executors.newSingleThreadExecutor())`
`Schedulers.from(Executors.newSingleThreadExecutor(), true)`
Both had the expected result.

`Schedulers.from(Executors.newSingleThreadScheduledExecutor(), false)`
`Schedulers.from(Executors.newSingleThreadScheduledExecutor(), true)`
Both ended with an interruption.
